dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/filter/FilterChainBuilder.java

```diff
@@ -144,12 +126,116 @@ public String toString() {
-            extends FilterChainNode<T, TYPE, FILTER> implements ClusterInvoker<T> {
+        extends FilterChainNode<T, TYPE, FILTER> implements ClusterInvoker<T> {
         public ClusterFilterChainNode(TYPE originalInvoker, Invoker<T> nextNode, FILTER filter) {
             super(originalInvoker, nextNode, filter);
         }
 
 
+        @Override
+        public URL getRegistryUrl() {
+            return getOriginalInvoker().getRegistryUrl();
+        }
+
+        @Override
+        public Directory<T> getDirectory() {
+            return getOriginalInvoker().getDirectory();
+        }
+
+        @Override
+        public boolean isDestroyed() {
+            return getOriginalInvoker().isDestroyed();
+        }
+    }
+
+    class CallbackRegistrationInvoker<T, FILTER extends BaseFilter> implements Invoker<T> {
+        static final Logger LOGGER = LoggerFactory.getLogger(CallbackRegistrationInvoker.class);
+        private final Invoker<T> filterInvoker;
+        private final List<FILTER> filters;
+
+        public CallbackRegistrationInvoker(Invoker<T> filterInvoker, List<FILTER> filters) {
+            this.filterInvoker = filterInvoker;
+            this.filters = filters;
+        }
+
+        @Override
+        public Result invoke(Invocation invocation) throws RpcException {
+            Result asyncResult = filterInvoker.invoke(invocation);
+            asyncResult.whenCompleteWithContext((r, t) -> {
+                for (int i = filters.size() - 1; i >= 0; i--) {
+                    FILTER filter = filters.get(i);
+                    try {
+                        if (filter instanceof ListenableFilter) {
+                            ListenableFilter listenableFilter = ((ListenableFilter) filter);
+                            Filter.Listener listener = listenableFilter.listener(invocation);
+                            try {
+                                if (listener != null) {
+                                    if (t == null) {
+                                        listener.onResponse(r, filterInvoker, invocation);
+                                    } else {
+                                        listener.onError(t, filterInvoker, invocation);
+                                    }
+                                }
+                            } finally {
+                                listenableFilter.removeListener(invocation);
+                            }
+                        } else if (filter instanceof FILTER.Listener) {
+                            FILTER.Listener listener = (FILTER.Listener) filter;
+                            if (t == null) {
+                                listener.onResponse(r, filterInvoker, invocation);
+                            } else {
+                                listener.onError(t, filterInvoker, invocation);
+                            }
+                        }
+                    } catch (Throwable filterThrowable) {
```

Comment:
Calling FILTER.Listener after all filters invoked may cause influence semantics of invokers, such as this may effect some environment clean up operation in finally block.

Example:
```java
public class DemoFilter implements Filter, Filter.Listener {

    @Override
    public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcException {
        try {
            return invoker.invoke(invocation);
        } finally {
            // action 1
        }
    }

    @Override
    public void onResponse(Result appResponse, Invoker<?> invoker, Invocation invocation) {
            // action 2
    }

    @Override
    public void onError(Throwable t, Invoker<?> invoker, Invocation invocation) {
            // action 2
    }
}
```

After this PR change, action 1 will happends before action 2.

Reply:
> After this PR change, action 1 will happends before action 2.

After thinking about this issue, I came up with two possible problematic scenarios:

```java
        @Override
        public Result invoke(Invocation invocation) throws RpcException {
            Result asyncResult = filterInvoker.invoke(invocation);
            asyncResult.whenCompleteWithContext((r, t) -> {
                for (int i = filters.size() - 1; i >= 0; i--) {
                    FILTER filter = filters.get(i);
```

## 1 The execution order of action1 and action2 in the same Fitler instance

`Result asyncResult = filterInvoker.invoke(invocation);` happens before `asyncResult.whenCompleteWithContext`(the onResponse callback registration). This sequence can make sure  that`action1` will always happen before `action2`, which I think is right and consistent with the previous behavior. 

## 2 The sequence of action1 and action2 among different filters

For filter chain like `Pxoxy -> Filter1 -> Filter2 -> Invoker`

Filter1's finally block(action1) can be executed before Filter2's onRespose callback(action2), which can be different from the previous behavior when making synchronous calls.  Typical example is `ClassLoaderFilter` which tries to set context that will be used by all the following filters and stacks, that is, all filters that set context across filters may be affected.

But in asynchronous scenarios, this behavior is also not guaranteed before.

Reply:
To conclude, there may still have some problems in certain scenarios for this `onResponse` model we are having now, but I don't think it's brought by changes in this patch.